metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesProvider.scala

```diff
@@ -75,32 +86,63 @@ final class TestSuitesProvider(
 
   override def onChange(docs: TextDocuments, file: AbsolutePath): Unit =
     docs.documents.headOption.foreach {
-      case doc if isEnabled => updateTestCases((file, doc))
+      case doc if isExplorerEnabled => updateTestCases((file, doc))
       case _ => ()
     }
 
   override def onDelete(file: AbsolutePath): Unit = {
     val removed = index.remove(file)
-    val removeEvents = removed
-      .groupBy(_.buildTarget)
-      .map { case (buildTarget, entries) =>
-        BuildTargetUpdate(
-          buildTarget,
-          entries.map(_.suiteDetails.asRemoveEvent),
-        )
-      }
-      .toList
+    if (isExplorerEnabled) {
+      val removeEvents = removed
+        .groupBy(_.buildTarget)
+        .map { case (buildTarget, entries) =>
+          BuildTargetUpdate(
+            buildTarget,
+            entries.map(_.suiteDetails.asRemoveEvent),
+          )
+        }
+        .toList
 
-    updateClientIfNonEmpty(removeEvents)
+      updateClientIfNonEmpty(removeEvents)
+    }
   }
 
   override def reset(): Unit = ()
 
+  override def codeLenses(
+      textDocumentWithPath: TextDocumentWithPath
+  ): Seq[l.CodeLens] = {
+    val path = textDocumentWithPath.filePath
+    for {
+      target <- buildTargets.inverseSources(path).toList
+      cases <- getTestCasesForPath(path, None)
+      lens <- cases.events.asScala.collect { case AddTestCases(fqn, _, cases) =>
+        cases.asScala.flatMap { entry =>
+          val c = ScalaTestSuiteSelection(fqn, List(entry.name).asJava)
+          val params = new b.DebugSessionParams(
+            List(target).asJava,
+            b.DebugSessionParamsDataKind.SCALA_TEST_SUITES_SELECTION,
+            ScalaTestSuites(List(c).asJava, Nil.asJava, Nil.asJava).toJson,
+          )
+          def lens(name: String, cmd: BaseCommand) = new l.CodeLens(
+            entry.location.getRange(),
+            new l.Command(name, cmd.id, List[Object](params).asJava),
+            null,
+          )
+          List(
+            lens("test", ClientCommands.StartRunSession),
+            lens("debug test", ClientCommands.StartDebugSession),
```

Comment:
```suggestion
            lens("test case", ClientCommands.StartRunSession),
            lens("debug test case", ClientCommands.StartDebugSession),
```
to differentiate from running full suite?